File: apps/python-sdk/firecrawl/v2/methods/aio/agent.py
from typing import Any, Dict, List, Literal, Optional, Union
import asyncio

from ...types import (
    AgentExchangeOptions,
    AgentListResponse,
    AgentResponse,
    AgentSnapshotResponse,
    AgentThreadResponse,
    AgentTraceResponse,
    AgentWebhookConfig,
    AuditMetadata,
    ThreatProtectionOptions,
)
from ...utils.error_handler import handle_response_error
from ...utils.http_client_async import AsyncHttpClient
from ...utils.validation import _normalize_schema


def _prepare_agent_request(
    urls: Optional[List[str]],
    *,
    prompt: str,
    schema: Optional[Any] = None,
    integration: Optional[str] = None,
    max_credits: Optional[int] = None,
    strict_constrain_to_urls: Optional[bool] = None,
    model: Optional[Literal["spark-1-pro", "spark-1-mini", "spark-2"]] = None,
    effort: Optional[Literal["low", "medium", "high"]] = None,
    webhook: Optional[Union[str, AgentWebhookConfig]] = None,
    threat_protection: Optional[ThreatProtectionOptions] = None,
    audit_metadata: Optional[AuditMetadata] = None,
    thread_id: Optional[str] = None,
    mode: Optional[Literal["extract", "chat"]] = None,
    exchange: Optional[Union[AgentExchangeOptions, Dict[str, Any]]] = None,
) -> Dict[str, Any]:
    body: Dict[str, Any] = {}
    if urls is not None:
        body["urls"] = urls
    body["prompt"] = prompt
    if schema is not None:
        normalized_schema = _normalize_schema(schema)
        if normalized_schema is not None:
            body["schema"] = normalized_schema
        else:
            raise ValueError(
                f"Invalid schema type: {type(schema).__name__}. "
                "Schema must be a dict, Pydantic BaseModel class, or Pydantic model instance."
            )
    if integration is not None and str(integration).strip():
        body["integration"] = str(integration).strip()
    if max_credits is not None and max_credits > 0:
        body["maxCredits"] = max_credits
    if strict_constrain_to_urls is not None and strict_constrain_to_urls:
        body["strictConstrainToURLs"] = strict_constrain_to_urls
    if model is not None:
        body["model"] = model
    if effort is not None:
        body["effort"] = effort
    if webhook is not None:
        if isinstance(webhook, str):
            body["webhook"] = webhook
        else:
            body["webhook"] = webhook.model_dump(exclude_none=True)
    if threat_protection is not None:
        body["threatProtection"] = threat_protection.model_dump(
            by_alias=True, exclude_none=True
        )
    if audit_metadata is not None:
        body["auditMetadata"] = audit_metadata.model_dump()
    if thread_id is not None:
        body["threadId"] = thread_id
    if mode is not None:
        body["mode"] = mode
    if exchange is not None:
        body["exchange"] = (
            exchange
            if isinstance(exchange, dict)
            else exchange.model_dump(by_alias=True, exclude_none=True)
        )
    return body


def _normalize_agent_response_payload(payload: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(payload)
    if "expiresAt" in out and "expires_at" not in out:
        out["expires_at"] = out["expiresAt"]
    if "creditsUsed" in out and "credits_used" not in out:
        out["credits_used"] = out["creditsUsed"]
    return out


async def start_agent(
    client: AsyncHttpClient,
    urls: Optional[List[str]],
    *,
    prompt: str,
    schema: Optional[Any] = None,
    integration: Optional[str] = None,
    max_credits: Optional[int] = None,
    strict_constrain_to_urls: Optional[bool] = None,
    model: Optional[Literal["spark-1-pro", "spark-1-mini", "spark-2"]] = None,
    effort: Optional[Literal["low", "medium", "high"]] = None,
    webhook: Optional[Union[str, AgentWebhookConfig]] = None,
    threat_protection: Optional[ThreatProtectionOptions] = None,
    audit_metadata: Optional[AuditMetadata] = None,
    thread_id: Optional[str] = None,
    mode: Optional[Literal["extract", "chat"]] = None,
    exchange: Optional[Union[AgentExchangeOptions, Dict[str, Any]]] = None,
) -> AgentResponse:
    body = _prepare_agent_request(
        urls,
        prompt=prompt,
        schema=schema,
        integration=integration,
        max_credits=max_credits,
        strict_constrain_to_urls=strict_constrain_to_urls,
        model=model,
        effort=effort,
        webhook=webhook,
        threat_protection=threat_protection,
        audit_metadata=audit_metadata,
        thread_id=thread_id,
        mode=mode,
        exchange=exchange,
    )
    resp = await client.post("/v2/agent", body)
    if not resp.ok:
        handle_response_error(resp, "agent")
    payload = _normalize_agent_response_payload(resp.json())
    return AgentResponse(**payload)


async def get_agent_status(client: AsyncHttpClient, job_id: str) -> AgentResponse:
    resp = await client.get(f"/v2/agent/{job_id}")
    if not resp.ok:
        handle_response_error(resp, "agent-status")
    payload = _normalize_agent_response_payload(resp.json())
    return AgentResponse(**payload)


async def list_agents(
    client: AsyncHttpClient,
    *,
    before: Optional[int] = None,
) -> AgentListResponse:
    """List agent runs, most recent first.

    Pages are fixed at 20 runs. To fetch the next page, pass the `before`
    value from the previous page's `next` URL. This method does not
    auto-paginate.

    Args:
        client: Async HTTP client instance
        before: Only return agent runs created before this unix ms timestamp

    Returns:
        AgentListResponse with the list of agent runs and optional next URL
    """
    endpoint = "/v2/agent"
    if before is not None:
        endpoint = f"{endpoint}?before={before}"
    resp = await client.get(endpoint)
    if not resp.ok:
        handle_response_error(resp, "list agents")
    return AgentListResponse(**resp.json())


async def wait_agent(
    client: AsyncHttpClient,
    job_id: str,
    *,
    poll_interval: int = 2,
    timeout: Optional[int] = None,
) -> AgentResponse:
    start_ts = asyncio.get_event_loop().time()
    while True:
        status = await get_agent_status(client, job_id)
        if status.status in ("completed", "failed", "cancelled"):
            return status
        if timeout is not None and (asyncio.get_event_loop().time() - start_ts) > timeout:
            return status
        await asyncio.sleep(max(1, poll_interval))


async def agent(
    client: AsyncHttpClient,
    urls: Optional[List[str]],
    *,
    prompt: str,
    schema: Optional[Any] = None,
    integration: Optional[str] = None,
    poll_interval: int = 2,
    timeout: Optional[int] = None,
    max_credits: Optional[int] = None,
    strict_constrain_to_urls: Optional[bool] = None,
    model: Optional[Literal["spark-1-pro", "spark-1-mini", "spark-2"]] = None,
    effort: Optional[Literal["low", "medium", "high"]] = None,
    webhook: Optional[Union[str, AgentWebhookConfig]] = None,
    threat_protection: Optional[ThreatProtectionOptions] = None,
    audit_metadata: Optional[AuditMetadata] = None,
    thread_id: Optional[str] = None,
    mode: Optional[Literal["extract", "chat"]] = None,
    exchange: Optional[Union[AgentExchangeOptions, Dict[str, Any]]] = None,
) -> AgentResponse:
    started = await start_agent(
        client,
        urls,
        prompt=prompt,
        schema=schema,
        integration=integration,
        max_credits=max_credits,
        strict_constrain_to_urls=strict_constrain_to_urls,
        model=model,
        effort=effort,
        webhook=webhook,
        threat_protection=threat_protection,
        audit_metadata=audit_metadata,
        thread_id=thread_id,
        mode=mode,
        exchange=exchange,
    )
    job_id = getattr(started, "id", None)
    if not job_id:
        return started
    return await wait_agent(client, job_id, poll_interval=poll_interval, timeout=timeout)


async def get_agent_trace(
    client: AsyncHttpClient,
    job_id: str,
    *,
    live_view: bool = False,
) -> AgentTraceResponse:
    """Get the execution trace of an agent job (spark-2 runs only).

    Args:
        client: Async HTTP client instance
        job_id: ID of the agent job
        live_view: Also include currently active browser sessions with live view URLs
    """
    endpoint = f"/v2/agent/{job_id}/trace"
    if live_view:
        endpoint += "?liveView=true"
    resp = await client.get(endpoint)
    if not resp.ok:
        handle_response_error(resp, "agent-trace")
    return AgentTraceResponse(**resp.json())


async def get_agent_thread(
    client: AsyncHttpClient,
    thread_id: str,
    *,
    include_data: bool = False,
) -> AgentThreadResponse:
    """Get a thread and its runs, oldest turn first.

    Args:
        client: Async HTTP client instance
        thread_id: Thread ID, as returned by start_agent or get_agent_status
        include_data: Inline each succeeded run's data
    """
    endpoint = f"/v2/agent/threads/{thread_id}"
    if include_data:
        endpoint += "?includeData=true"
    resp = await client.get(endpoint)
    if not resp.ok:
        handle_response_error(resp, "agent-thread")
    return AgentThreadResponse(**resp.json())


async def get_agent_snapshot(
    client: AsyncHttpClient,
    job_id: str,
    snapshot_id: str,
) -> AgentSnapshotResponse:
    """Get the full content of an artifact snapshot referenced by a trace event.

    Args:
        client: Async HTTP client instance
        job_id: ID of the agent job
        snapshot_id: Snapshot ID from an artifact.updated trace event
    """
    resp = await client.get(f"/v2/agent/{job_id}/snapshots/{snapshot_id}")
    if not resp.ok:
        handle_response_error(resp, "agent-snapshot")
    return AgentSnapshotResponse(**resp.json())


async def cancel_agent(client: AsyncHttpClient, job_id: str) -> bool:
    """
    Cancel a running agent job.

    Args:
        client: Async HTTP client instance
        job_id: ID of the agent job to cancel

    Returns:
        bool: True if the agent was cancelled, False otherwise

    Raises:
        Exception: If the cancellation fails
    """
    resp = await client.delete(f"/v2/agent/{job_id}")
    if not resp.ok:
        handle_response_error(resp, "cancel agent")
    return resp.json().get("success", False)

File: apps/js-sdk/firecrawl/src/v2/methods/agent.ts
import { type AgentEffort, type AgentExchangeOptions, type AgentListOptions, type AgentListResponse, type AgentMode, type AgentResponse, type AgentSnapshotResponse, type AgentStatusResponse, type AgentThreadResponse, type AgentTraceResponse, type AgentWebhookConfig, type AuditMetadata, type ThreatProtectionOptions } from "../types";
import { HttpClient } from "../utils/httpClient";
import { normalizeAxiosError, throwForBadResponse } from "../utils/errorHandler";
import { isZodSchema, zodSchemaToJsonSchema } from "../../utils/zodSchemaToJson";
import type { ZodTypeAny } from "zod";

function prepareAgentPayload(args: {
  urls?: string[];
  prompt: string;
  schema?: Record<string, unknown> | ZodTypeAny;
  integration?: string;
  origin?: string;
  maxCredits?: number;
  strictConstrainToURLs?: boolean;
  model?: "spark-1-pro" | "spark-1-mini" | "spark-2";
  effort?: AgentEffort;
  webhook?: string | AgentWebhookConfig;
  threatProtection?: ThreatProtectionOptions;
  auditMetadata?: AuditMetadata;
  threadId?: string;
  mode?: AgentMode;
  exchange?: AgentExchangeOptions;
}): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (args.urls) body.urls = args.urls;
  body.prompt = args.prompt;
  if (args.schema != null) {
    body.schema = isZodSchema(args.schema) ? zodSchemaToJsonSchema(args.schema) : args.schema;
  }
  if (args.integration && args.integration.trim()) body.integration = args.integration.trim();
  if (args.origin) body.origin = args.origin;
  if (args.maxCredits !== null && args.maxCredits !== undefined) body.maxCredits = args.maxCredits;
  if (args.strictConstrainToURLs !== null && args.strictConstrainToURLs !== undefined) body.strictConstrainToURLs = args.strictConstrainToURLs;
  if (args.model !== null && args.model !== undefined) body.model = args.model;
  if (args.effort !== null && args.effort !== undefined) body.effort = args.effort;
  if (args.webhook != null) body.webhook = args.webhook;
  if (args.threatProtection != null)
    body.threatProtection = args.threatProtection;
  if (args.auditMetadata != null)
    body.auditMetadata = args.auditMetadata;
  if (args.threadId != null) body.threadId = args.threadId;
  if (args.mode != null) body.mode = args.mode;
  if (args.exchange != null) body.exchange = args.exchange;
  return body;
}

export async function startAgent(http: HttpClient, args: Parameters<typeof prepareAgentPayload>[0]): Promise<AgentResponse> {
  const payload = prepareAgentPayload(args);
  try {
    const res = await http.post<AgentResponse>("/v2/agent", payload);
    if (res.status !== 200) throwForBadResponse(res, "agent");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "agent");
    throw err;
  }
}

export async function getAgentStatus(http: HttpClient, jobId: string): Promise<AgentStatusResponse> {
  try {
    const res = await http.get<AgentStatusResponse>(`/v2/agent/${jobId}`);
    if (res.status !== 200) throwForBadResponse(res, "agent status");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "agent status");
    throw err;
  }
}

export async function listAgents(http: HttpClient, options?: AgentListOptions): Promise<AgentListResponse> {
  try {
    const query = options?.before !== undefined ? `?before=${options.before}` : "";
    const res = await http.get<AgentListResponse>(`/v2/agent${query}`);
    if (res.status !== 200) throwForBadResponse(res, "list agents");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "list agents");
    throw err;
  }
}

export async function waitAgent(
  http: HttpClient,
  jobId: string,
  pollInterval = 2,
  timeout?: number
): Promise<AgentStatusResponse> {
  const start = Date.now();
  while (true) {
    const status = await getAgentStatus(http, jobId);
    if (["completed", "failed", "cancelled"].includes(status.status || "")) return status;
    if (timeout != null && Date.now() - start > timeout * 1000) return status;
    await new Promise((r) => setTimeout(r, Math.max(1000, pollInterval * 1000)));
  }
}

export async function agent(
  http: HttpClient,
  args: Parameters<typeof prepareAgentPayload>[0] & { pollInterval?: number; timeout?: number }
): Promise<AgentStatusResponse> {
  const started = await startAgent(http, args);
  const jobId = started.id;
  if (!jobId) return started as unknown as AgentStatusResponse;
  return waitAgent(http, jobId, args.pollInterval ?? 2, args.timeout);
}

export async function getAgentTrace(
  http: HttpClient,
  jobId: string,
  options?: { liveView?: boolean }
): Promise<AgentTraceResponse> {
  try {
    const query = options?.liveView ? "?liveView=true" : "";
    const res = await http.get<AgentTraceResponse>(`/v2/agent/${jobId}/trace${query}`);
    if (res.status !== 200) throwForBadResponse(res, "agent trace");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "agent trace");
    throw err;
  }
}

export async function getAgentSnapshot(
  http: HttpClient,
  jobId: string,
  snapshotId: string
): Promise<AgentSnapshotResponse> {
  try {
    const res = await http.get<AgentSnapshotResponse>(`/v2/agent/${jobId}/snapshots/${snapshotId}`);
    if (res.status !== 200) throwForBadResponse(res, "agent snapshot");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "agent snapshot");
    throw err;
  }
}

export async function getAgentThread(
  http: HttpClient,
  threadId: string,
  options?: { includeData?: boolean }
): Promise<AgentThreadResponse> {
  try {
    const query = options?.includeData ? "?includeData=true" : "";
    const res = await http.get<AgentThreadResponse>(`/v2/agent/threads/${threadId}${query}`);
    if (res.status !== 200) throwForBadResponse(res, "agent thread");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "agent thread");
    throw err;
  }
}

export async function cancelAgent(http: HttpClient, jobId: string): Promise<boolean> {
  try {
    const res = await http.delete<{ success: boolean }>(`/v2/agent/${jobId}`);
    if (res.status !== 200) throwForBadResponse(res, "cancel agent");
    return res.data?.success === true;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "cancel agent");
    throw err;
  }
}
